File: OOP in Python/OOP.py
# First watch complelte Vedio! Then you can understand the whole code easily
# OOP in Python
# ----------------
# TO map with real world scenarios, we stated using object in code. This is called object Oriented Programming


# CLass & Object in Python
# Before starting OOP we need to understand Objects & Classes 
# Objects => Are entities in the real world (pen, mouse etc.)
# Class => Is like a blueprint of these entities
# Syntax of class
# class Name {
#     properities
#     .
#     .
#     .
#     methods()    
# }; 
# --------------------------
# Class is a blur print for creating objects.

# Let's try to create class:
class Student:
    name = "Muhammad Habib"

# Let's try to create object(instance)
s1 = Student()
print(s1.name)

s2 = Student()
print(s1.name)

# You can created many objects as you want w.r.t class

___________________________________________________________________________________________________________


# __init__ Function
# Constructor
# All classes have a function called _init_(), which is always executed when the class is being initiated.

# Let's some code for constructor
class Student:

    College_Name = "Govt. COllege"
    name = "Anonymous"  # CLass attribut < Obj attribute

    # Default contructors
    def __init__(self):
        print("I'm default constructor")

    #Parameterized Constructors
    def __init__(self, name, Marks):   # This is a parameter for object and we can give many parameters as we want
        print(self)  # Object
        self.name = name    # Obj attribute > Class attribute
        self.Marks = Marks    
        print("Adding new student in database......")

s1 = Student("Abdul Munnam", 72)  # Output => Adding new student in database......
print(s1) # Output => Object ; as we can see self and s1 has same output
print(s1.name, s1.Marks)  # Abdul Munnam    72
print(Student.College_Name)
# 2nd Method to access class attribute
print(s1.College_Name)

s2 = Student("Ali Ahmad", 80)  # Output => Adding new student in database......
print(s2) # Output => Object ; as we can see self and s1 has same output
print(s2.name, s2.Marks)  # Ali Ahmad    80
print(Student.College_Name)
# 2nd Method to access class attribute
print(s2.College_Name)

# Constructor always avoke by itself
# The self parameter is a reference to the current instance of the class, and is used to acess variables that belong to the class



_____________________________________________________________________________________________________________


# Methods => Are functions  that belongs to objects.

class Students:

    def __init__(self, name, Marks):  
        self.name = name
        self.Marks = Marks   

    def welcome(self):
        print("Welcome Students")

    def get_marks(self):
        return self.Marks
    

s1 = Student("Habib", 97)
s1.welcome()
print(s1.get_marks())



_____________________________________________________________________________________________________________

# Static Methods => Methods that don't use the self parameter (Work at class level)

class Student:
    @staticmethod     # decorator
    def college():
        print("ABC College")

# decorator allow us to wrap another function in order to extend the behavior of the wrapped function, without permanene modifying it.



_________________________________________________________________________________________________________________

# Important Concepts

# 1. Abstraction => Hiding the implementation details of a class and only showing the essential features to the user.

class car:
    def __init__(self):
        self.accelerator = False
        self.brk = False
        self.clutch = False

    def start(self):
        self.clutch = True
        self.accelerator = True
        print("Car start......")


car1 = car()
car1.start()

# 2. Encapsulation => Wrapping data and function into a single unit (Object).

____________________________________________________________________________________________________________________

# del keyword => It is used to delete object properties or object itself

# for example if we have object named s1 and we want to delete it so we can delete it in the following way

del s1   # It delete object itself

# If object has property like name so we can delete it

del s1.name   # delete naem property


_____________________________________________________________________________________________________________________

# Private attributes & methods

# Conceptual Implementation in python
# Private attribute & methods are meant to be used only within the class and are not accessible from outside the class

class Account:
    def __init__(self, acc_no, acc_pass):
        self.acc_no = acc_no
        self.__acc_pass = acc_pass   # By adding "__" our attribute become private

    def method(self):
        print(self.__acc_pass)  # Here we can access password


acc1 = Account("1234", "asffd")
print(acc1.acc_no)
print(acc1.__acc_pass)   # We Can not acces password here because it become provate and it's only accessible in class

# Let's take an example

class Person:
    __name = "Abdullah"

    def __hello(self):
        print("Hello User!!!!!!!")

    def welcome(self):
        print(self.__hello)

p1 = Person()
print(p1.welcome()) # This is OK scenario
print(p1.__hello()) # But this is not OK seen. It will give error

___________________________________________________________________________________________________________________

# 3 Inheritance => When one class derives the properties & methods of another calss is called....

class Car:
    @staticmethod
    def start():
        print("Car started.....")

    @staticmethod
    def stop():
        print("Car Stopped........")

class ToyotaCar(Car):
    def __init__(self, name):
        self.name = name


car1 = ToyotaCar("Fortuner")
car2 = ToyotaCar("Prius")

print(car1.name)
print(car1.start())


# There are three types of Inheritance
# 1. Single Inheritance    this type of Inheritance discuss above
# 2. Multi-level Inheritance
# 3. Multiple Inheritance


# Multi-level Inheritance
class Car:
    @staticmethod
    def start():
        print("Car started.....")

    @staticmethod
    def stop():
        print("Car Stopped........")

class ToyotaCar(Car):
    def __init__(self, brand):
        self.brand = brand


class Fortuner(ToyotaCar):
    def __init__(self, type):
        self.type = type


car1 = Fortuner("Diesel")
car1.start()



# Multiple Inheritance
class A:
    varA = "WElcome to class A"

class B:
    varB = "WElcome to class B"

class C(A, B):
    varA = "WElcome to class C"


c1 = C()
print(c1.varC)
print(c1.varB)
print(c1.varA)

_____________________________________________________________________________________________________________________

File: OOP in JavaScript/OOP.js
// Before starting Object-Oriented Programming in JS, you need in to understand one importand thing that: "javaScript is not pure object oriented programming language" However C++, python, Java etc. are pure object oriented programming

// So Now let's get started!
// As we know OOP base on two important things that are "Objects" and "Classes" these are the main structure that OOP base on it

// What is class?? => Class is just like a blue print for any object creation
// Syntax => class Name {........}

// Let's take one example to create class

`class Student {
    constructor () {
        this.name = "Abdul Munnam"
        this.age = 20
        this.height = "6 Foot"
    }
}`

/* Well, you have created class but what is the "contructor", "this" in class? 
Ok, Let me explain
constructor: it is just like birth certificate for class 
- but the question is that why we used it??
- We used it in that case when we assign same values to many objects
let's explain it with the help of example */ 

class Student {
    constructor(name, age, syllabus) {
        this.name = name;
        this.age = age ;
        this.syllabus = syllabus;
    }
}

// here "this" refers to the object that we will create like s2, s3 ....

let s2 = new Student("Abdul Munnam", 20, "Full Book")
let s3 = new Student("Ali Ahmad", 19, "Full Book")

console.log(s2, s3);

/* There are two types of constructors 
- Parameterized
- Non-Parameterized 
here we used parameterized constructor */


// Let's discuss Inheritance in OOP

// You can inherit one class properties the another class. Example is given below
class Car {
    constructor (model) {
        this.model = model;
    }
}

class Fortuner extends Car {
    constructor (color, CarNum) {
        super("A12")
        this.color = color
        this.CarNum = CarNum

    }
}


let c1 = new Fortuner("Black", "01")
console.log(c1.model);

// Note: super() method can only called once



// Now discuss about getter and setter function

class Person {
    constructor(firstName, lastName) {
      this._firstName = firstName;
      this._lastName = lastName;
    }
  
    get fullName() {
      return `${this._firstName} ${this._lastName}`;  
  
    }
  
    set fullName(value) {
      const parts = value.split(' ');
      this._firstName = parts[0];
      this._lastName = parts[1];
  
    }
  }
  
  const person = new Person('Abdul', 'Munnam');  // By using this method getter function doing because we give seperate first and last name
  console.log(person.fullName); // Output: Abdul Munnam 
  
// What if we give one string as an input? to handle this case we used setter function

person.fullName = 'Muhammad Habib';
console.log(person.fullName);




/* last concept is static method
static method allow you to access any method properties within class youo can not access it's properties through objects */

class Student {
    constructor(name,age) {
        this.name = name;
        this.age = age;
    }
    static  marks(mark) {
        this.marks = mark
    }
}

let s4 = new Student("Saad",18)
console.log(s4.marks);